### DISCUSSION TITLE
Fixing "Livewire is not defined" on various hosting environments

### DISCUSSION THREAD

**Discussion by stevebauman** [livewire/livewire#3089] [7. Bugs (archived)]
Please bear with me -- I know this is a giant wall of text, but I felt it is necessary to address the concerns that I will mention with absolute clarity.

### Problem:

The `/livewire/livewire.js` Laravel route causes issues with certain hosting environments that (by default) route all *file based* requests to the hosts filesystem. This is due to the `.js` file suffix -- as there is not *actually* a file residing in the users `public/` directory (without publishing Livewire's assets).

### Description:

Currently in all versions of Livewire, the recommended approach is to leave `livewire.js` file unpublished, and to simply use the `@livewireScripts` Blade directive to include the `<script>` tag to the `/livewire/livewire.js` route that is registered inside the `LivewireServiceProvider`:

https://github.com/livewire/livewire/blob/8663232c198ef12964b62559e9bb2023eb86701d/src/LivewireServiceProvider.php#L129-L130

Ex:
```blade
<!-- Directive: -->
@livewireScripts

<!-- Produces: -->
<script src="/livewire/livewire.js?id=936e5d0fb0b76b631ba7" data-turbo-eval="false" data-turbolinks-eval="false"></script>
```

The problem with this, is that most Nginx configs that I've seen, and the built-in PHP development server, will not proxy the `.js` file request **through** Laravel itself. It will instead attempt to look for an actual file on the server in the `public/livewire/livewire.js` path, but return a 404, because the server says the file does not actually exist.

For example, the below screenshots were taken using the built-in PHP web server (`cd public && php -S localhost:8000`):

---

#### Console:
![Screen Shot 2021-01-11 at 5 58 29 PM](https://user-images.githubusercontent.com/6421846/104247929-9e8dea80-5436-11eb-8a24-c26ad3dce637.png)

---

#### PHP Server Error:
![Screen Shot 2021-01-11 at 5 58 53 PM](https://user-images.githubusercontent.com/6421846/104247967-ac437000-5436-11eb-85b4-ea51aa78ce55.png)

---

#### PHP Server Console Error:
<img width="850" alt="Screen Shot 2021-01-11 at 6 04 20 PM" src="https://user-images.githubusercontent.com/6421846/104248409-6f2bad80-5437-11eb-9eb2-b3f0a9be7d52.png">

---

We can resolve this specific issue for **all** hosting environments by simply removing the `.js` suffix on the route (if the user has not published their assets). The request will then be treated as a normal application request, and the proper `livewire.js` file contents will be returned from the web server. Livewire does not need to specify the `.js` extension for the script to be properly loaded.

However, while this will actually make Livewire *work*, we will also need to adjust how Livewire routes messages from components.

Livewire will attempt to communicate with components with the URL `/livewire/message/{name}` route defined here:

https://github.com/livewire/livewire/blob/8663232c198ef12964b62559e9bb2023eb86701d/src/LivewireServiceProvider.php#L132-L133

Though this introduces the same issue as above, since the Livewire attempts to use the URL convention below, which includes an unescaped period (`.`), which the server will infer as a file request:

```
http://localhost/livewire/message/{component}.{method}
```

![Screen Shot 2021-01-11 at 6 18 18 PM](https://user-images.githubusercontent.com/6421846/104249428-62a85480-5439-11eb-9885-9238b42fed63.png)

I can submit a PR to resolve these issues -- but before I do that, I wanted to get another set of eyes on this in case I'm not seeing something that could cause BC breaks.

Please let me know if any further clarity is needed, or if you'd like me to just submit the PR and you can accept/deny it there 👍 

Thanks for your time! ❤️ 

### Related issues:
- https://github.com/livewire/livewire/issues/794
- https://github.com/laravel/jetstream/issues/112
- https://github.com/laravel/jetstream/issues/236 (and more on the Jetstream repo)

### Exact steps to reproduce:

1. `laravel new livewire-test --jet`
2. Select `livewire` (`0`)
3. Run `npm i && npm run dev`
3. Change `DB_CONNECTION` to `sqlite` in `.env`
4. Run `php artisan migrate`
5. Run `cd public && php -S localhost:8000`
6. Open the browser console and attempt to use Jetstream normally

### Environment Context:

- Livewire version: All
- Laravel version: N/A
- Alpine version: N/A
- Browser: All

**Comment by ju5t** [comment]
Wouldn't `try_files` do the trick? You should be able to add it before any fastcgi parameters. If the file isn't there, it would just continue.

**Comment by stevebauman** [comment]
@ju5t Yes -- but that's not the point of this issue. We can resolve this completely for any hosting environment by simply structuring the URI's differently.

As new developers attempt to use Livewire, we shouldn't require them to modify their hosting configuration to simply be able to use it.

**Comment by ju5t** [comment]
@stevebauman just trying to help. I am still curious if your pretty urls work, they probably will, but it's not what I expected. In the [docs](https://laravel.com/docs/8.x/deployment#nginx) the two go hand in hand.

In the end I have nothing against a PR. My flimsy personal preference would be to have the suffix there. But hey, in the end it doesn't matter as long as it works.

**Comment by stevebauman** [comment]
>@stevebauman just trying to help. 

@ju5t Oh for sure! I apologize if my tone came off in a negative way -- I didn't mean for that to be the case ❤️ 

> My flimsy personal preference would be to have the suffix there. But hey, in the end it doesn't matter as long as it works.

Maybe we could have it default to no `.js` extension to provide the most support, and allow devs to customize the route themselves? Such as using a static variable to store the registered route URL:

```php
Livewire::setScriptURL('/livewire/livewire.js');

// Or:

Livewire::loadScriptsUsing(function ($controller, $action)) {
    Route::get('/livewire/livewire.js', [$controller, $action]);
}
```

Thoughts?

**Comment by ju5t** [comment]
No worries!

> Maybe we could have it default to no .js extension to provide the most support, and allow devs to customize the route themselves? 

I can only come up with one argument for leaving the default as is: people who use a CDN or some other caching mechanism might have caching rules set up based on an extension instead of the content type. 

If they cache everything on `*.js`, that would break.

> Thoughts?

As much as I like your suggestion, I would probably go for a setting myself (on/off).

<sup>Ps. I happen to own a hosting company. Most of our servers run Apache. I did check our Nginx configurations after your post and it wouldn't work on our servers either. This needs some work on our side too. On it!</sup>

**Comment by calebporzio** [comment]
Mmmm thanks for the thorough issue and offer to PR @stevebauman.

I agree that this needs to be addressed if running PHP's built-in server causes a breakage.

My main concern is what @ju5t said. Will this cause caching issues on the front-end without the .js suffix?

@stevebauman (and others), if you would be willing to do some digging on the implications of changing to a non ".js" default path for the JS assets (mostly with respect to browser caching) that would be GREAT! I'd love a PR if your findings provide a clear path forward.
 
In the meantime: I'd welcome a PR to escape the periods in the /livewire/message/... path - that seems like a good call to me that wouldn't be a breaking change (I don't think at least).

Thanks a bunch!

**Comment by stevebauman** [comment]
Happy to help @calebporzio!

> My main concern is what @ju5t said. Will this cause caching issues on the front-end without the .js suffix?

Yeah this is absolutely a concern -- let me do some testing with my own fork of livewire with the suggested fixes above to see if the response is still cached properly.

> In the meantime: I'd welcome a PR to escape the periods in the /livewire/message/... path - that seems like a good call to me that wouldn't be a breaking change (I don't think at least).

I'll also give this a shot. I think we should be able to simply `encodeURIComponent()` on the JS side while not touching the backend, as Laravel should decode the URI and perform the request normally.

I'll update here if I gain any ground! 👍

**Comment by calebporzio** [comment]
Awesome! Thanks for tackling this @stevebauman !

**Comment by RomkaLTU** [comment]
This problem exists on Forge installs, just stumbled on this. 
It's happened after the last update: 2.3.17 
File actually accessible via browser but gives 404 headers so browser not trying to read it. 

I actually forced to publish assets manually, because non of the fixes worked for me. And as I already mentioned this problem introduced after 2.3.17, because I have other projects on the same server and they working fine.

**Comment by gaetandezeiraud** [comment]
Same problem on my side. `php artisan livewire:publish --assets` works.

**Comment by devutoo** [comment]
Hi,
same problem here. On my local server (XAMPP on a Windows machine) it works fine.
After deploying my project on a stage server (Ubuntu, Apache) I get the Livewire not defined error and landed here after some searching.

And now Livewire is found, but not only the first Livewire components are shown. Every interaction, that requires Livewire to communicate with the backend, produces an overlay and a 404 for the component.

There is a long thread in the Laravel Livewire forum, where someone could reproduce that behavior even with the simple counter example (https://forum.laravel-livewire.com/t/404-error-on-any-post-request/2277). I.e., Livewire will load after publishing the assets, but there seems to be a communication problem with the backend.

I hope, that hints someone of you smarter people in the right direction. Maybe there is a server-sided workaround, that helps to solve the issue.

Fingers crossed.

**Comment by ju5t** [comment]
@devutoo in the thread that you copied in it seems like it was an issue with their Nginx configuration. This happens to be in the Laravel docs https://laravel.com/docs/7.x/installation#pretty-urls. It also includes an example for Apache. This assumes you're using mod_php and not something like php-fpm.

Have you tried that?

**Reply by sonofgod1** [reply]
This worked for me, thanks: I am on a shared hosting (hostgator) and the problem occurred when publishing my application, livewire not defined. But I went to the laravel link and copied the content of .htaccess php to my .htaccess and the problem was removed.

**Comment by devutoo** [comment]
Thank you @ju5t . 
I am using mod_php. And I changed the .htaccess according to the docs section, that you proposed. Unfortunately it didn't solve the problem.

Edit: I tried it on a fresh VM with Ubuntu Server. Via putty I created a new Laravel project, installed Livewire and recreated the counter example. It didn't work neither.

Edit2: I tried different Livewire versions (from 2.2.0 to latest) and it didn't matter.

**Comment by sirshaun** [comment]
Hi @stevebauman, did you manage to create a pr for this and has it been merged? I am facing the same problems as outlined in your original post with the built-in php dev server, and I am losing my mind trying to solve this.

> I'll also give this a shot. I think we should be able to simply `encodeURIComponent()` on the JS side while not touching the backend, as Laravel should decode the URI and perform the request normally.

**Comment by devutoo** [comment]
> 
> 
> Thank you @ju5t .
> I am using mod_php. And I changed the .htaccess according to the docs section, that you proposed. Unfortunately it didn't solve the problem.
> 
> Edit: I tried it on a fresh VM with Ubuntu Server. Via putty I created a new Laravel project, installed Livewire and recreated the counter example. It didn't work neither.
> 
> Edit2: I tried different Livewire versions (from 2.2.0 to latest) and it didn't matter.

SOLVED/WORKAROUND: In my case the public/.htaccess file of my Laravel installation was not recognized. I had to change a setting of the apache2.conf (apache config file on Ubuntu). 
I edited the 
```
<Directory /var/www/>
AllowOverride None
</Directory\>
```
to
```
<Directory /var/www/>
AllowOverride All
</Directory>
```
As a result the standard Laravel .htaccess can finally do its job and Livewire works as expected. (ref.: https://stackoverflow.com/questions/18740419/how-to-set-allowoverride-all).

**Comment by mosabalrojpany** [comment]
the sulotion is remove prefix from routeserviceprovider and add it to route group
`Route::group(['middleware' => 'auth','prefix' => 'cp'], function()
{
    Route::get('logout',[LoginController::class, 'logout'])->name('admin.logout');
    Route::resource('index',LoginController::class)->names([
        'index' => 'admin.index',]);

    Route::get('prusase_order/print/{id}',[PrushaseOrderController::class, 'print'])->name('admin.prusase_order.print');
    Route::post('prusase_order/Merchant',[PrushaseOrderController::class, 'Merchant'])->name('admin.prusase_order.Merchant');
    Route::resource('prusase_order',PrushaseOrderController::class)->names([
        'index' => 'admin.prusase_order',
        'create' => 'admin.prusase_order.create',
        'store' => 'admin.prusase_order.store',
        'show' => 'admin.prusase_order.show',
        'edit' => 'admin.prusase_order.edit',
        'update' => 'admin.prusase_order.update',
        'destroy' => 'admin.prusase_order.destroy'
        ]);
});`
do not do this 
`    public function boot()
    {
        $this->configureRateLimiting();

        $this->routes(function () {
            Route::prefix('api')
                ->middleware('api')
                ->namespace($this->namespace)
                ->group(base_path('routes/api.php'));

            Route::prefix('admin')
                ->middleware('web')
                ->namespace($this->namespace)
                ->group(base_path('routes/web.php'));
        });
    }`
remove prefix('admin')

**Comment by mrkheradmand** [comment]
Hello guys,
I have same issue when uploaded project to host, I could fix it by 
1. setting the APP_URL in ENV file 
2. updating file (vendor\livewire\livewire\config\livewire.php) with this: 'asset_url' => env('APP_URL', 'http://localhost'),
this saved me alot.
Hope it will fix yours.

**Reply by acharseth** [reply]
And you also have to publish the assets, right?

**Comment by acharseth** [comment]
With my first Livewire attempt I got this issue. Not with Livewire.js but with a message: Failed to load resource...: https://mydomain/myapp/current/livewire/message/containerloc.form
So is this still an issue or have I done something wrong?
How do I fix this?

**Reply by acharseth** [reply]
I had a mistake in my config/livewire.php. My `app_url` was pointing to root instead of `public`.

**Comment by hrsa** [comment]
I had the same problem on nginx, and solved the issue by creating a symlink :
` ln -s ../vendor/livewire/livewire/dist ./public/livewire`
Maybe this could be a solution ?

**Reply by pmeyers** [reply]
I just ran into this issue running Laravel Herd and this fixed it for me.

**Comment by carrilhorafael57** [comment]
I just ran into this issue. I am a newbie in laravel/livewire/filament building an application and on localhost it runs just fine.

I tried to host it in one server to test it live, however, I am getting this error.

My host domain is e.g. example.com and my laravel/livewire/filament and the project login route to the filament dashboard panel is example.com/test/public/admin.

My livewire simply do not know how to redirect the request when I sign to my dashboard.

How can I fix this? I am hosting it like this because I am sharing the vm with another website for testing.

Thanks!

**Comment by CrazyBoy49z** [comment]
> I just ran into this issue. I am a newbie in laravel/livewire/filament building an application and on localhost it runs just fine.
> 
> I tried to host it in one server to test it live, however, I am getting this error.
> 
> My host domain is e.g. example.com and my laravel/livewire/filament and the project login route to the filament dashboard panel is example.com/test/public/admin.
> 
> My livewire simply do not know how to redirect the request when I sign to my dashboard.
> 
> How can I fix this? I am hosting it like this because I am sharing the vm with another website for testing.
> 
> Thanks!

Edit nginx config

```
location ^~ /livewire {
        try_files $uri $uri/ /index.php?$query_string;
    }
```
before     
```
location ~* ^.+\.(css|js|jpg...
```
